### DISCUSSION TITLE
Reasoning behind default m and u params?

### DISCUSSION THREAD

**Discussion by NickCrews** [moj-analytical-services/splink#1426] [General]
Hey @RobinL, thanks for the help here.

I'm wondering about the reasoning behind https://github.com/moj-analytical-services/splink/blob/4935a6a48e3fb32003687d748969b96ba3648ce9/splink/comparison_level.py#L80-L101

If I try with 4 levels, then I see
```
print(_default_m_values(4))
print(_default_u_values(4))
[0.01666666666666668, 0.01666666666666668, 0.01666666666666668, 0.95]
[0.5333333333333338, 0.03333333333333336, 0.002083333333333335, 0.000927734375]
```

This looks like we are assuming that the first level is the ELSE and corresponds to a non-match (low m, high u), and the last level is the best match (high m, low u). But there is nothing that keeps users from giving their levels in this order at all, and so I could see this optimism leading to bad results if the user doesn't follow this assumption. Is there a problem with giving all the levels equal weights and not making any assumptions? eg `[.25, .25, .25, .25]`?

**Comment by RobinL** [comment]
Yeah, the decisions there are fairly arbitrary but your description of them is pretty much what I was thinking when I designed them.

One thing to watch out for here is that the values need to be 'ballpark' right for EM to start iterating in the right direction.

One way of seeing this is the algorithm (or to be more specific the likelihood function) is 'symmetric' in the sense it can just as easily iterate to exactly the wrong answer as exactly the right one (labelling matches as non matches, and vice versa, specifically `1-match_probability` rather than `match_probability` ).  By 'just as easily' i mean the log likelihood  of 'exactly the wrong answer' is the same as the right answer.  That also means it's kind of inherently unstable and will only iterate to the right answer if the starting parameters are somewhat right.

This is particularly the case you run  [estimate_parameters_using_expectation_maximisation](https://github.com/moj-analytical-services/splink/blob/8bf2600356db053ccbc4b1573465108a5d0a4fe2/splink/linker.py#L1481) to estimate all parameters (not just m).

I guess my logic is that generally (but i accept not always) the first comparison level will be an exact match.  Unless your data quality is terrible, amongst true matches, fields usually match.  So it's not unreasonable to set the m value as 0.95 for the exact match.

But if you have better priors, particularly for the m values, you should use them

The starting u values are even more arbitrary, but are based on the idea that the 'else' condition is more likely than the 'exact match'  (again, I'm assuming the first level is exact match).  But I think this matters less because  we have a fairly reliable estimation techique for `u` values (estimate_u_using_random_sampling) so we  never really use  the starting values in practice.

Find I guess another thing worth considering is that if you use some of our default out of the box comparisons, then these starting values are fairly reasonable starting points, or at least possibly difficult to improve upon

**Reply by NickCrews** [reply]
Thanks for the explanation, that all makes sense. I am asking because I'm trying to re-implement this in a port of splink that uses Ibis, and now I'm trying to understand the EM part.

I think I am going to go with
- Set u values using random sampling, just like splink.
- Set all m values to be the same, at 1/n_levels

Therefore
1. The u values will lean the EM to iterate in the correct direction
2. we don't have to make any assumptions about the order of the levels

I could also imagine setting the initial m values to be based off the u values, so that high u values cause low m values and vice versa.
